administration/src/util/qrcode.ts

```diff
@@ -1,37 +1,184 @@
 import {
-  EncodeHintType,
+  BitArray,
   IllegalStateException,
-  QRCodeDecoderErrorCorrectionLevel as ErrorCorrectionLevel,
+  QRCodeByteMatrix,
+  QRCodeDecoderErrorCorrectionLevel,
   QRCodeEncoder,
   QRCodeEncoderQRCode as QRCode,
+  QRCodeMatrixUtil,
+  QRCodeMode,
+  QRCodeVersion,
 } from '@zxing/library'
 import { PDFPage, rgb } from 'pdf-lib'
+import ECBlocks from '@zxing/library/esm/core/qrcode/decoder/ECBlocks'
+import MatrixUtil from '@zxing/library/esm/core/qrcode/encoder/MatrixUtil'
+import MaskUtil from '@zxing/library/esm/core/qrcode/encoder/MaskUtil'
 
-const DEFAULT_QUIET_ZONE_SIZE = 10
+const DEFAULT_QUIET_ZONE_SIZE = 4 // pt
+
+// Level 8 with EC of M gives 152 bytes
+// Level 7 with EC of L gives 154 bytes
+// Level 8 seems appropriate right now.
+export const DEFAULT_VERSION: QRCodeVersion = QRCodeVersion.getVersionForNumber(8)
+// From EC L to M we have a double of EC capability: 7% -> 15%
+export const DEFAULT_ERROR_CORRECTION: QRCodeDecoderErrorCorrectionLevel = QRCodeDecoderErrorCorrectionLevel.M
+
+function calculateBitsNeeded(
+  mode: QRCodeMode,
+  headerBitsSize: number,
+  dataBitsSize: number,
+  version: QRCodeVersion
+): number {
+  return headerBitsSize + mode.getCharacterCountBits(version) + dataBitsSize
+}
+
+function willFit(
+  mode: QRCodeMode,
+  headerBitsSize: number,
+  dataBitsSize: number,
+  version: QRCodeVersion,
+  ecLevel: QRCodeDecoderErrorCorrectionLevel
+): boolean {
```

Comment:
Can you add a link that explains these calculations?

Reply:
This code is copied from [Encoder.ts](https://github.com/zxing-js/library/blob/master/src/core/qrcode/encoder/Encoder.ts). I'll try to clarify this better by adding a comment.

I do not want to change their code to keep compatibility and avoid mistakes.